hibernate-core/src/main/java/org/hibernate/metamodel/model/domain/internal/AnyMappingDomainTypeImpl.java

```diff
@@ -19,10 +19,15 @@
 public class AnyMappingDomainTypeImpl<T> implements AnyMappingDomainType<T> {
 	private final AnyType anyType;
 	private final JavaType<T> baseJtd;
+	private final AnyDiscriminatorDomainTypeImpl<?> anyDiscriminatorType;
 
 	public AnyMappingDomainTypeImpl(AnyType anyType, JavaType<T> baseJtd) {
 		this.anyType = anyType;
 		this.baseJtd = baseJtd;
+		final MetaType discriminatorType = (MetaType) anyType.getDiscriminatorType();
+		anyDiscriminatorType = new AnyDiscriminatorDomainTypeImpl<>(
```

Comment:
`AnyDiscriminatorDomainTypeImpl` does not implement `BasicType`.

By the way not sure here how to use the `ConvertedBasicTypeImpl`

Reply:
Oh ok, `AnyDiscriminatorDomainTypeImpl` looked so similar to `DiscriminatorType` that I thought you copied that. It was just an idea, not sure if this actually works, but it would be instantiated like this:

```
anyDiscriminatorType = new ConvertedBasicTypeImpl<>(
  null, // no name
  discriminatorType.getBaseType().getJdbcType(),
  new AnyDiscriminatorConverter<>( discriminatorType )
);
```

where `AnyDiscriminatorConverter` would roughly look like this:

```
public class AnyDiscriminatorConverter<T> implements BasicValueConverter<T, Object> {
	private final MetaType modelPart;

	public AnyDiscriminatorDomainTypeImpl(MetaType modelPart) {
		this.modelPart = modelPart;
	}

	@Override
	public T toDomainValue(Object discriminatorValue) {
		if ( discriminatorValue == null ) {
			return null;
		}
		final String entityName = modelPart.getDiscriminatorValuesToEntityNameMap().get( discriminatorValue );
		if ( entityName == null ) {
			throw new HibernateException( "Unable to resolve discriminator value [" + discriminatorValue + "] to entity name" );
		}
		final EntityPersister entityPersister = persister.getFactory()
				.getRuntimeMetamodels()
				.getMappingMetamodel()
				.getEntityDescriptor( entityName );
		//noinspection unchecked
		return entityPersister.getRepresentationStrategy().getMode() == RepresentationMode.POJO
				? (T) entityPersister.getJavaType().getJavaTypeClass()
				: (T) entityName;
	}

	@Override
	public Object toRelationalValue(T domainForm) {
		if ( domainForm == null ) {
			return null;
		}

		if ( domainForm instanceof Class ) {
			return modelPart.getEntityNameToDiscriminatorValueMap().get( ( (Class) domainForm ).getName() );
		}
		else {
			return modelPart.getEntityNameToDiscriminatorValueMap().get( (String) domainForm );
		}
	}

	@Override
	public JavaType<T> getDomainJavaType() {
		return (JavaType<T>) ClassJavaType.INSTANCE;
	}

	@Override
	public JavaType<Object> getRelationalJavaType() {
		return underlyingType.getExpressibleJavaType();
	}
}
```

I kept the `T` type parameter for the domain type, but from what I understand, it always has to be a `Class`.